hw/mcu/native/src/hal_adc.c

```diff
@@ -0,0 +1,315 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <hal/hal_adc.h>
+#include <hal/hal_adc_int.h>
+#include <mcu/mcu_devid.h>
+#include <mcu/mcu_hal.h>
+
+/* forwards for the const structure below */
+static int native_adc_random_read(struct hal_adc_device_s *padc);
+static int native_adc_random_get_resolution(struct hal_adc_device_s *padc);
+static int native_adc_random_get_reference_mvolts(struct hal_adc_device_s *padc);
+
+/* This ADC had two channels that return 8-bit 
+ * random numbers */
+const struct hal_adc_funcs_s random_adc_funcs = {
+    .hadc_get_reference_mvolts = &native_adc_random_get_reference_mvolts,
+    .hadc_get_resolution = &native_adc_random_get_resolution,
+    .hadc_read = &native_adc_random_read,
+};
+
+struct hal_adc_device_s * 
+native_adc_random_create(enum McuDeviceDescriptor devid) 
+{
+    struct hal_adc_device_s *padc = NULL;
+    if((devid == MCU_ADC_CHANNEL_0) || (devid == MCU_ADC_CHANNEL_1)) {
+        padc = malloc(sizeof(struct hal_adc_device_s));
+        
+        if(padc) {
+            padc->driver_api = &random_adc_funcs;
+            
+            /* no other internal state to set */
+        }
+    }
+    return padc;
+}
+
+static int 
+native_adc_random_get_resolution(struct hal_adc_device_s *padc) 
+{
+    if(padc && padc->driver_api == &random_adc_funcs) {
+         return 8;        
+    }
+    return -1;
+}
+
+static int 
+native_adc_random_get_reference_mvolts(struct hal_adc_device_s *padc) 
+{
+    if(padc && padc->driver_api == &random_adc_funcs) {
+        return 5000;    
+    }
+    return -1;
+}
+
+static int 
+native_adc_random_read(struct hal_adc_device_s *padc) 
+{
+    if(padc && padc->driver_api == &random_adc_funcs) {
+    return rand() & 0xff;
+    }
+    return -1;    
+}
+
+/* forwards for the const structure below */
+static int native_adc_mmm_read(struct hal_adc_device_s *padc);
+static int native_adc_mmm_get_resolution(struct hal_adc_device_s *padc);
+static int native_adc_mmm_get_reference_mvolts(struct hal_adc_device_s *padc);
+
+/* This ADC had two channels that return 8-bit 
+ * random numbers */
+const struct hal_adc_funcs_s mmm_adc_funcs = {
+    .hadc_get_reference_mvolts = &native_adc_mmm_get_reference_mvolts,
+    .hadc_get_resolution = &native_adc_mmm_get_resolution,
+    .hadc_read = &native_adc_mmm_read,
+};
+
+/* This driver needs to keep a bit of state */
+struct mmm_adc_device_s {
+    struct hal_adc_device_s parent;
+    enum McuDeviceDescriptor devid;   
+};
+
+struct hal_adc_device_s*
+native_adc_mmm_create(enum McuDeviceDescriptor devid) 
+{
+    struct mmm_adc_device_s *padc = NULL;
+    if(        (devid == MCU_ADC_CHANNEL_2) 
```

Comment:
standard is if ((devid == MCU_ADC_CHANNEL_2) ||
                      (devid == MCU_ADC_CHANNEL_3) ||
matching the inner parentheses with the start of the line, but including the || statement at the end of the line.